doc/developer/design/20231201_catalog_migration_to_persist.md

```diff
@@ -0,0 +1,164 @@
+
+The catalog durably stores a fencing token called the epoch. All catalog writers and readers also
+store an in-memory epoch. When a new writer or reader connects to the catalog they store the
+previous epoch plus one in memory and persist this new value durably. Every time a reader or write
+performs a read or write, they compare their in-memory token with the persisted token, and if
+there's a difference, the operation fails. As a result, whenever a reader/write increments the epoch
+durably, they are effectively fencing out all previous readers/writers.
+
+### Catalog Open
+
+Opening a catalog involves the following two steps:
+
+1. Increment the persist epoch.
+2. Either initialize the persisted catalog state if it's uninitialized, or migrate the data format
+   if there are any migrations, otherwise do nothing.
+
+### Migrate From Stash to Persist
+
+Below are the steps to migrate from the stash to persist.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. If the stash tombstone is `true`, then we're done.
+4. Read stash catalog snapshot.
+5. Write catalog snapshot to persist.
+6. Set the stash tombstone to `true`.
+
+### Rollback From Persist to Stash
+
+Below are the steps to rollback from persist to the stash.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. If the stash tombstone is `false` or doesn't exist, then we're done.
+4. Read persist catalog snapshot.
+5. Write catalog snapshot to stash.
+6. Set the stash tombstone to `false`.
+
+NOTE: Steps (5) and (6) can be done as a single write as a performance optimization.
+
+### States
+
+Below is a table describing all possible states of the stash tombstone and what they
+indicate about where the source of truth catalog data is located.
+
+| Stash Tombstone | Source of Truth | Explanation                                                                 |
+|-----------------|-----------------|-----------------------------------------------------------------------------|
+| None            | Stash           | Migration has never fully completed.                                        |
+| Some(true)      | Persist         | Migration has completed successfully or rollback crashed before completing. |
+| Some(false)     | Stash           | Rollback has completed successfully or migration crashed before completing. |
+
+### State Transitions
+
+Below is a state transition diagram that shows all possible states of the stash tombstone and how
+the states transition. Each node has the value of the stash tombstone. Each edge indicates what
+steps from the algorithm above triggers that specific state transition. Next to each node is either
+`Stash` if the stash is the source of truth, or `Persist` if persist is the source of truth.
+
+![state-transitions](./static/catalog_migration_to_persist/catalog_migration_single_tombstone.png)
+
+## Alternatives
+
+Below is a similar algorithm that uses two tombstone values, one in persist and one in the stash.
```

Comment:
I also prefer the proposed design over this alternative because I like having a single place to check for which one to use